In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/used-car-dataset-ford-and-mercedes/hyundi.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/vauxhall.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/cclass.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/merc.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/unclean focus.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/focus.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/unclean cclass.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/audi.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/toyota.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/vw.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/ford.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/skoda.csv
/kaggle/input/used-car-dataset-ford-and-mercedes/bmw.csv


In [2]:
import numpy as np
import pandas as pd

In [3]:
audi=pd.read_csv('/kaggle/input/used-car-dataset-ford-and-mercedes/audi.csv')


In [4]:
audi.isnull().sum()

model           0
year            0
price           0
transmission    0
mileage         0
fuelType        0
tax             0
mpg             0
engineSize      0
dtype: int64

In [5]:
from sklearn.preprocessing import LabelEncoder


In [6]:
lc=LabelEncoder()
audi.model=lc.fit_transform(audi.model)
audi.transmission=lc.fit_transform(audi.transmission)
audi.fuelType=lc.fit_transform(audi.fuelType)


In [7]:
X=audi.drop('price',axis=1)
y=audi.price

In [8]:
columns=X[['model','transmission','fuelType']]
columns.columns

Index(['model', 'transmission', 'fuelType'], dtype='object')

In [9]:
l=pd.get_dummies(data=columns,columns=columns.columns)
l.head(3)

model_0  model_1  model_2  model_3  model_4  model_5  model_6  model_7  \
0        1        0        0        0        0        0        0        0   
1        0        0        0        0        0        1        0        0   
2        1        0        0        0        0        0        0        0   

   model_8  model_9  ...  model_22  model_23  model_24  model_25  \
0        0        0  ...         0         0         0         0   
1        0        0  ...         0         0         0         0   
2        0        0  ...         0         0         0         0   

   transmission_0  transmission_1  transmission_2  fuelType_0  fuelType_1  \
0               0               1               0           0           0   
1               1               0               0           1           0   
2               0               1               0           0           0   

   fuelType_2  
0           1  
1           0  
2           1  

[3 rows x 32 columns]

In [10]:
audi['model'].unique()

array([ 0,  5,  3,  2,  9, 10,  4, 20,  8,  6, 25, 11, 17, 14,  7, 12, 15,
       16, 13, 23, 22, 24, 19, 21,  1, 18])

In [11]:
X.drop(['model','transmission','fuelType'],axis=1,inplace=True)

In [12]:
X_=pd.concat((X,l),axis=1)

In [13]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from xgboost import XGBRegressor
import lightgbm as lgb
from catboost import CatBoostRegressor
from sklearn.metrics import r2_score


In [14]:
X_train, X_test, y_train, y_test = train_test_split(X_,y,test_size=.25)

In [15]:
LiR=LinearRegression()
LiR.fit(X_train,y_train)
pred1=LiR.predict(X_test)

DcT=DecisionTreeRegressor()
DcT.fit(X_train,y_train)
pred2=DcT.predict(X_test)


svr=SVR()
svr.fit(X_train,y_train)
pred3=svr.predict(X_test)


naive=GaussianNB()
naive.fit(X_train,y_train)
pred4=naive.predict(X_test)


knn=KNeighborsRegressor()
knn.fit(X_train,y_train)
pred5=knn.predict(X_test)


RFR=RandomForestRegressor()
RFR.fit(X_train,y_train)
pred6=RFR.predict(X_test)



gbr=GradientBoostingRegressor()
gbr.fit(X_train,y_train)
pred7=gbr.predict(X_test)


xgb=XGBRegressor()
xgb.fit(X_train,y_train)
pred8=xgb.predict(X_test)


lgb=lgb.LGBMRegressor()
lgb.fit(X_train,y_train)
pred9=lgb.predict(X_test)


cat=CatBoostRegressor()
cat.fit(X_train,y_train)
pred10=cat.predict(X_test)


Learning rate set to 0.056515
0:	learn: 11242.7959097	total: 67.3ms	remaining: 1m 7s
1:	learn: 10767.5424311	total: 78.9ms	remaining: 39.4s
2:	learn: 10335.0396801	total: 89.4ms	remaining: 29.7s
3:	learn: 9909.9922083	total: 101ms	remaining: 25.1s
4:	learn: 9517.5002353	total: 113ms	remaining: 22.4s
5:	learn: 9138.6211756	total: 124ms	remaining: 20.5s
6:	learn: 8791.9821267	total: 129ms	remaining: 18.3s
7:	learn: 8460.4381346	total: 132ms	remaining: 16.4s
8:	learn: 8150.7263521	total: 135ms	remaining: 14.9s
9:	learn: 7834.0129586	total: 139ms	remaining: 13.7s
10:	learn: 7546.0369228	total: 142ms	remaining: 12.7s
11:	learn: 7275.1584751	total: 144ms	remaining: 11.9s
12:	learn: 7043.6019653	total: 147ms	remaining: 11.2s
13:	learn: 6807.1885647	total: 151ms	remaining: 10.6s
14:	learn: 6583.0163927	total: 153ms	remaining: 10.1s
15:	learn: 6373.3969469	total: 157ms	remaining: 9.64s
16:	learn: 6175.3254602	total: 159ms	remaining: 9.22s
17:	learn: 5986.0967384	total: 162ms	remaining: 8.85s
18

In [16]:
print('r2 Score of LinearRegression Model: ', r2_score(y_test,pred1))
print('r2 Score of DecisionTreeRegressor Model: ', r2_score(y_test,pred2))
print('r2 Score of SVR Model: ', r2_score(y_test,pred3))
print('r2 Score of GaussianNB Model: ', r2_score(y_test,pred4))
print('r2 Score of KNeighborsRegressor Model: ', r2_score(y_test,pred5))
print('r2 Score of RandomForestRegressor Model: ', r2_score(y_test,pred6))
print('r2 Score of GradientBoostingRegressor Model: ', r2_score(y_test,pred7))
print('r2 Score of XGBRegressor Model: ', r2_score(y_test,pred8))
print('r2 Score of LGBMRegressor Model: ', r2_score(y_test,pred9))
print('r2 Score of CatBoostRegressor Model: ', r2_score(y_test,pred10))


r2 Score of LinearRegression Model:  0.8876833472193626
r2 Score of DecisionTreeRegressor Model:  0.9331626219103828
r2 Score of SVR Model:  0.027127480757369038
r2 Score of GaussianNB Model:  0.7219692412785759
r2 Score of KNeighborsRegressor Model:  0.48954152546194674
r2 Score of RandomForestRegressor Model:  0.9608356115209841
r2 Score of GradientBoostingRegressor Model:  0.9406476773294896
r2 Score of XGBRegressor Model:  0.9631091352554016
r2 Score of LGBMRegressor Model:  0.9600876254422189
r2 Score of CatBoostRegressor Model:  0.9664139636609774


# CatBoostRegressor Gave r2_score of 96.7 

## Applying Standard Scaler 😜

In [17]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

In [18]:
X_=scaler.fit_transform(X_)


In [19]:
X_train, X_test, y_train, y_test = train_test_split(X_,y,test_size=.25)

In [20]:
import lightgbm as lgb


In [21]:
LiR=LinearRegression()
LiR.fit(X_train,y_train)
pred1=LiR.predict(X_test)

DcT=DecisionTreeRegressor()
DcT.fit(X_train,y_train)
pred2=DcT.predict(X_test)


svr=SVR()
svr.fit(X_train,y_train)
pred3=svr.predict(X_test)


naive=GaussianNB()
naive.fit(X_train,y_train)
pred4=naive.predict(X_test)


knn=KNeighborsRegressor()
knn.fit(X_train,y_train)
pred5=knn.predict(X_test)


RFR=RandomForestRegressor()
RFR.fit(X_train,y_train)
pred6=RFR.predict(X_test)



gbr=GradientBoostingRegressor()
gbr.fit(X_train,y_train)
pred7=gbr.predict(X_test)


xgb=XGBRegressor()
xgb.fit(X_train,y_train)
pred8=xgb.predict(X_test)


lgb=lgb.LGBMRegressor()
lgb.fit(X_train,y_train)
pred9=lgb.predict(X_test)


cat=CatBoostRegressor()
cat.fit(X_train,y_train)
pred10=cat.predict(X_test)


Learning rate set to 0.056515
0:	learn: 11239.3888283	total: 3.02ms	remaining: 3.02s
1:	learn: 10775.0035276	total: 5.6ms	remaining: 2.79s
2:	learn: 10333.8519146	total: 8.11ms	remaining: 2.69s
3:	learn: 9916.2485686	total: 11.1ms	remaining: 2.75s
4:	learn: 9522.2727318	total: 13.5ms	remaining: 2.69s
5:	learn: 9159.5824686	total: 15.9ms	remaining: 2.64s
6:	learn: 8813.7022159	total: 19.2ms	remaining: 2.72s
7:	learn: 8468.2494069	total: 21.8ms	remaining: 2.7s
8:	learn: 8151.2151309	total: 24.3ms	remaining: 2.67s
9:	learn: 7867.6432680	total: 27.2ms	remaining: 2.69s
10:	learn: 7596.5884084	total: 30.2ms	remaining: 2.72s
11:	learn: 7334.8154542	total: 32.9ms	remaining: 2.71s
12:	learn: 7085.1709674	total: 36ms	remaining: 2.73s
13:	learn: 6865.0435418	total: 38.8ms	remaining: 2.73s
14:	learn: 6636.0998572	total: 41.7ms	remaining: 2.73s
15:	learn: 6430.3490276	total: 44.8ms	remaining: 2.76s
16:	learn: 6215.3069743	total: 48.1ms	remaining: 2.78s
17:	learn: 6021.4106186	total: 51.2ms	remainin

In [22]:
print('r2 Score of LinearRegression Model: ', r2_score(y_test,pred1))
print('r2 Score of DecisionTreeRegressor Model: ', r2_score(y_test,pred2))
print('r2 Score of SVR Model: ', r2_score(y_test,pred3))
print('r2 Score of GaussianNB Model: ', r2_score(y_test,pred4))
print('r2 Score of KNeighborsRegressor Model: ', r2_score(y_test,pred5))
print('r2 Score of RandomForestRegressor Model: ', r2_score(y_test,pred6))
print('r2 Score of GradientBoostingRegressor Model: ', r2_score(y_test,pred7))
print('r2 Score of XGBRegressor Model: ', r2_score(y_test,pred8))
print('r2 Score of LGBMRegressor Model: ', r2_score(y_test,pred9))
print('r2 Score of CatBoostRegressor Model: ', r2_score(y_test,pred10))


r2 Score of LinearRegression Model:  0.8873433124033102
r2 Score of DecisionTreeRegressor Model:  0.937158496403892
r2 Score of SVR Model:  -0.006379147949759023
r2 Score of GaussianNB Model:  0.7248504912178761
r2 Score of KNeighborsRegressor Model:  0.9244785001607294
r2 Score of RandomForestRegressor Model:  0.9581647215583069
r2 Score of GradientBoostingRegressor Model:  0.9396674335095292
r2 Score of XGBRegressor Model:  0.9585721022845372
r2 Score of LGBMRegressor Model:  0.9592868836954449
r2 Score of CatBoostRegressor Model:  0.9624907868893953


# Still CatBoost in on Top with 96.2% Accuracy

### Kindly Upvote if you liked ✌️